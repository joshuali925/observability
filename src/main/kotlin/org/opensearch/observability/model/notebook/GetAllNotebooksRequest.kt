/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 *
 */

package org.opensearch.observability.model.notebook

import org.opensearch.observability.ObservabilityPlugin.Companion.LOG_PREFIX
import org.opensearch.observability.model.RestTag.FROM_INDEX_FIELD
import org.opensearch.observability.model.RestTag.MAX_ITEMS_FIELD
import org.opensearch.observability.settings.PluginSettings
import org.opensearch.observability.util.logger
import org.opensearch.action.ActionRequest
import org.opensearch.action.ActionRequestValidationException
import org.opensearch.common.io.stream.StreamInput
import org.opensearch.common.io.stream.StreamOutput
import org.opensearch.common.xcontent.ToXContent
import org.opensearch.common.xcontent.ToXContentObject
import org.opensearch.common.xcontent.XContentBuilder
import org.opensearch.common.xcontent.XContentFactory
import org.opensearch.common.xcontent.XContentParser
import org.opensearch.common.xcontent.XContentParser.Token
import org.opensearch.common.xcontent.XContentParserUtils
import org.opensearch.commons.utils.STRING_READER
import org.opensearch.commons.utils.STRING_WRITER
import org.opensearch.commons.utils.enumReader
import org.opensearch.commons.utils.fieldIfNotNull
import org.opensearch.commons.utils.stringList
import org.opensearch.observability.model.RestTag.FILTER_PARAM_LIST_FIELD
import org.opensearch.observability.model.RestTag.OBJECT_ID_LIST_FIELD
import org.opensearch.observability.model.RestTag.SORT_FIELD_FIELD
import org.opensearch.observability.model.RestTag.SORT_ORDER_FIELD
import org.opensearch.search.sort.SortOrder
import java.io.IOException

/**
 * Get All notebooks info request
 * Data object created from GET request query params
 * <pre> JSON format
 * {@code
 * {
 *   "fromIndex":100,
 *   "maxItems":100
 * }
 * }</pre>
 */
internal class GetAllNotebooksRequest(
    val objectIds: Set<String>?,
    val fromIndex: Int,
    val maxItems: Int,
    val sortField: String?,
    val sortOrder: SortOrder?,
    val filterParams: Map<String, String>
) : ActionRequest(), ToXContentObject {

    companion object {
        private val log by logger(GetAllNotebooksRequest::class.java)

        /**
         * Parse the data from parser and create [GetAllNotebooksRequest] object
         * @param parser data referenced at parser
         * @return created [GetAllNotebooksRequest] object
         */
        fun parse(parser: XContentParser): GetAllNotebooksRequest {
            var objectIdList: Set<String> = setOf()
            var fromIndex = 0
            var maxItems = PluginSettings.defaultItemsQueryCount
            var sortField: String? = null
            var sortOrder: SortOrder? = null
            var filterParams: Map<String, String> = mapOf()

            XContentParserUtils.ensureExpectedToken(Token.START_OBJECT, parser.currentToken(), parser)
            while (Token.END_OBJECT != parser.nextToken()) {
                val fieldName = parser.currentName()
                parser.nextToken()
                when (fieldName) {
                    OBJECT_ID_LIST_FIELD -> objectIdList = parser.stringList().toSet()
                    FROM_INDEX_FIELD -> fromIndex = parser.intValue()
                    MAX_ITEMS_FIELD -> maxItems = parser.intValue()
                    SORT_FIELD_FIELD -> sortField = parser.text()
                    SORT_ORDER_FIELD -> sortOrder = SortOrder.fromString(parser.text())
                    FILTER_PARAM_LIST_FIELD -> filterParams = parser.mapStrings()
                    else -> {
                        parser.skipChildren()
                        log.info("$LOG_PREFIX:Skipping Unknown field $fieldName")
                    }
                }
            }
            return GetAllNotebooksRequest(objectIdList, fromIndex, maxItems, sortField, sortOrder, filterParams)
        }
    }

    /**
     * {@inheritDoc}
     */
    @Throws(IOException::class)
    constructor(input: StreamInput) : this(
        objectIds = input.readStringList().toSet(),
        fromIndex = input.readInt(),
        maxItems = input.readInt(),
        sortField = input.readOptionalString(),
        sortOrder = input.readOptionalWriteable(enumReader(SortOrder::class.java)),
        filterParams = input.readMap(STRING_READER, STRING_READER)
    )

    /**
     * {@inheritDoc}
     */
    @Throws(IOException::class)
    override fun writeTo(output: StreamOutput) {
        super.writeTo(output)
        output.writeStringCollection(objectIds)
        output.writeInt(fromIndex)
        output.writeInt(maxItems)
        output.writeOptionalString(sortField)
        output.writeOptionalWriteable(sortOrder)
        output.writeMap(filterParams, STRING_WRITER, STRING_WRITER)
    }

    /**
     * {@inheritDoc}
     */
    override fun validate(): ActionRequestValidationException? {
        return if (fromIndex < 0) {
            val exception = ActionRequestValidationException()
            exception.addValidationError("fromIndex should be grater than 0")
            exception
        } else if (maxItems <= 0) {
            val exception = ActionRequestValidationException()
            exception.addValidationError("maxItems should be grater than or equal to 0")
            exception
        } else {
            null
        }
    }

    /**
     * create XContentBuilder from this object using [XContentFactory.jsonBuilder()]
     * @param params XContent parameters
     * @return created XContentBuilder object
     */
    fun toXContent(params: ToXContent.Params = ToXContent.EMPTY_PARAMS): XContentBuilder? {
        return toXContent(XContentFactory.jsonBuilder(), params)
    }

    /**
     * {@inheritDoc}
     */
    override fun toXContent(builder: XContentBuilder?, params: ToXContent.Params?): XContentBuilder {
        return builder!!.startObject()
            .field(OBJECT_ID_LIST_FIELD, objectIds)
            .field(FROM_INDEX_FIELD, fromIndex)
            .field(MAX_ITEMS_FIELD, maxItems)
            .fieldIfNotNull(SORT_FIELD_FIELD, sortField)
            .fieldIfNotNull(SORT_ORDER_FIELD, sortOrder)
            .field(FILTER_PARAM_LIST_FIELD, filterParams)
            .endObject()
    }
}
