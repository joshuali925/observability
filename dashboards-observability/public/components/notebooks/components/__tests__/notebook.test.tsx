/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { fireEvent, render } from '@testing-library/react';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { HttpResponse } from '../../../../../../../src/core/public';
import httpClientMock from '../../../../../test/__mocks__/httpClientMock';
import { Notebook } from '../notebook';

jest.mock('../../../../../../../src/plugins/embeddable/public', () => ({
  ViewMode: {
    EDIT: 'edit',
    VIEW: 'view',
  },
}));

// @ts-ignore
global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () =>
      Promise.resolve({
        status: {
          statuses: [{ id: 'plugin:reportsDashboards' }],
        },
      }),
  })
);

describe('<Notebook /> spec', () => {
  configure({ adapter: new Adapter() });

  it('renders the empty component', () => {
    const setBreadcrumbs = jest.fn();
    const renameNotebook = jest.fn();
    const cloneNotebook = jest.fn();
    const deleteNotebook = jest.fn();
    const setToast = jest.fn();
    const location = jest.fn();
    const history = jest.fn() as any;
    history.replace = jest.fn();
    const utils = render(
      <Notebook
        openedNoteId="mock-id"
        DashboardContainerByValueRenderer={jest.fn()}
        http={httpClientMock}
        parentBreadcrumb={{ href: 'parent-href', text: 'parent-text' }}
        setBreadcrumbs={setBreadcrumbs}
        renameNotebook={renameNotebook}
        cloneNotebook={cloneNotebook}
        deleteNotebook={deleteNotebook}
        setToast={setToast}
        location={location}
        history={history}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('renders the component', () => {
    const setBreadcrumbs = jest.fn();
    const renameNotebook = jest.fn();
    const cloneNotebook = jest.fn();
    const deleteNotebook = jest.fn();
    const setToast = jest.fn();
    const location = jest.fn();
    const history = jest.fn() as any;
    history.replace = jest.fn();
    httpClientMock.get = jest.fn(() => Promise.resolve((null as unknown) as HttpResponse));
    const utils = render(
      <Notebook
        openedNoteId="mock-id"
        DashboardContainerByValueRenderer={jest.fn()}
        http={httpClientMock}
        parentBreadcrumb={{ href: 'parent-href', text: 'parent-text' }}
        setBreadcrumbs={setBreadcrumbs}
        renameNotebook={renameNotebook}
        cloneNotebook={cloneNotebook}
        deleteNotebook={deleteNotebook}
        setToast={setToast}
        location={location}
        history={history}
      />
    );
    expect(utils.container.firstChild).toMatchSnapshot();
  });
});
