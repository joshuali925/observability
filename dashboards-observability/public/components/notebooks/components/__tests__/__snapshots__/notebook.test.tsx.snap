// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notebook /> spec renders the component 1`] = `
<div
  style="float: left; width: 100%; max-width: 1500px;"
>
  <div
    class="euiPage"
  >
    <div
      class="euiPageBody"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--justifyContentFlexEnd euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem"
        />
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        />
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        />
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiPopover euiPopover--anchorDownCenter euiPopover--withTitle"
          >
            <div
              class="euiPopover__anchor"
            >
              <button
                class="euiButton euiButton--primary"
                type="button"
              >
                <span
                  class="euiButtonContent euiButtonContent--iconRight euiButton__content"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                  <span
                    class="euiButton__text"
                  >
                    Paragraph actions
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        />
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiPopover euiPopover--anchorDownCenter euiPopover--withTitle"
          >
            <div
              class="euiPopover__anchor"
            >
              <button
                class="euiButton euiButton--primary"
                type="button"
              >
                <span
                  class="euiButtonContent euiButtonContent--iconRight euiButton__content"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiIcon-isLoading euiButtonContent__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                  <span
                    class="euiButton__text"
                  >
                    Notebook actions
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--s"
      />
      <h1
        class="euiTitle euiTitle--large"
      />
      <div
        class="euiSpacer euiSpacer--m"
      />
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsFlexStart euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiText euiText--medium"
          >
            <div>
              <p>
                Created 
                <br />
                 
                Invalid date
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        style="float: left; width: 100%; max-width: 1130px; margin-top: 20px;"
      >
        <div
          class="euiPanel euiPanel--paddingMedium"
        >
          <div
            class="euiSpacer euiSpacer--xxl"
          />
          <div
            class="euiText euiText--medium"
          >
            <div
              class="euiTextAlign euiTextAlign--center"
            >
              <h2>
                No paragraphs
              </h2>
              <div
                class="euiText euiText--medium"
              >
                Add a paragraph to compose your document or story. Notebooks now support two types of input:
              </div>
            </div>
          </div>
          <div
            class="euiSpacer euiSpacer--xl"
          />
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--justifyContentSpaceEvenly euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrow2"
            />
            <div
              class="euiFlexItem euiFlexItem--flexGrow3"
            >
              <div
                class="euiCard euiCard--paddingMedium euiCard--centerAligned euiCard--hasIcon"
              >
                <div
                  class="euiCard__top"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--xxLarge euiIcon-isLoading euiCard__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </div>
                <div
                  class="euiCard__content"
                >
                  <span
                    class="euiTitle euiTitle--small euiCard__title"
                    id="random_html_idTitle"
                  >
                    Code block
                  </span>
                  <div
                    class="euiText euiText--small euiCard__description"
                    id="random_html_idDescription"
                  >
                    <p>
                      Write contents directly using markdown, SQL or PPL.
                    </p>
                  </div>
                </div>
                <div
                  class="euiCard__footer"
                >
                  <button
                    class="euiButton euiButton--primary"
                    style="margin-bottom: 17px;"
                    type="button"
                  >
                    <span
                      class="euiButtonContent euiButton__content"
                    >
                      <span
                        class="euiButton__text"
                      >
                        Add code block
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrow3"
            >
              <div
                class="euiCard euiCard--paddingMedium euiCard--centerAligned euiCard--hasIcon"
              >
                <div
                  class="euiCard__top"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--xxLarge euiIcon-isLoading euiCard__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  />
                </div>
                <div
                  class="euiCard__content"
                >
                  <span
                    class="euiTitle euiTitle--small euiCard__title"
                    id="random_html_idTitle"
                  >
                    OpenSearch Dashboards visualization
                  </span>
                  <div
                    class="euiText euiText--small euiCard__description"
                    id="random_html_idDescription"
                  >
                    <p>
                      Import OpenSearch Dashboards visualizations to the notes.
                    </p>
                  </div>
                </div>
                <div
                  class="euiCard__footer"
                >
                  <button
                    class="euiButton euiButton--primary"
                    style="margin-bottom: 17px;"
                    type="button"
                  >
                    <span
                      class="euiButtonContent euiButton__content"
                    >
                      <span
                        class="euiButton__text"
                      >
                        Add visualization
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrow2"
            />
          </div>
          <div
            class="euiSpacer euiSpacer--xxl"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
